code/model/model.py

```diff
@@ -0,0 +1,227 @@
+                "Invalid decoder_type"
+            )
+
+        #ladder
+        self.top_down_layers = []
+        self.bottom_up_layers = []
+
+        #get a list contain bottom up layers,which used to generate z_mu_q_d and z_var_q_d
+        for i in range(len(self.d_size)):
+            if i == 0:
+                self.bottom_up_layers.append(MLP(in_size=self.ladder_input_size,layer_size=self.d_size[i],out_size=self.z_size[i]))
+            else:
+                self.bottom_up_layers.append(MLP(in_size=self.d_size[i-1],layer_size=self.d_size[i],out_size=self.z_size[i]))
+
+        #get a list contain top down layers,which used to generate z_mu_p and z_var_p
+        for i in range(len(self.z_reverse_size)-1):
+            self.top_down_layers.append(MLP(in_size=self.z_reverse_size[i],layer_size=self.z2z_layer_size[i],out_size=self.z_reverse_size[i+1]))
+
+    def bottom_up(self,input):
+        '''
+        Do the bottom up step and get z_mu_q_d and z_var_q_d
+
+        :param: input of ladder part,size=(batch_size * ladder_input_size)
+        :return:two lists: z_mu_q_d ,z_var_q_d
+        '''
+        z_mu_q_d = []
+        z_var_q_d = []
+        for i in range(len(self.d_size)):
+            if i == 0:
+                nn, mu_q_d, var_q_d = self.bottom_up_layers[i](input)
+                z_mu_q_d.append(mu_q_d)
+                z_var_q_d.append(var_q_d)
+            else:
+                nn, mu_q_d, var_q_d = self.bottom_up_layers[i](nn)
+                z_mu_q_d.append(mu_q_d)
+                z_var_q_d.append(var_q_d)
+        return z_mu_q_d, z_var_q_d
+
+    def top_down(self, z_sample, z_mu_p, z_var_p):
+        '''
+        Do top down step and get z_mu_p,z_var_p, also return samples of each level z
+
+        :param z_sample: only have the samples of top z
+        :param z_mu_p: only have the z_mu_p of top z (mu = 0)
+        :param z_var_p: only have the z_var_p of top z (var = 1)
+        :return three list :z_mu_p,z_var_p,z_sample
+        '''
+        for i in range(len(self.z_reverse_size) - 1):
+            _, mu_p, var_p = self.top_down_layers[i](z_sample[i])
+            z_sample.append(self.sample_z(mu_p, var_p))
+            z_mu_p.append(mu_p)
+            z_var_p.append(var_p)
+        return list(reversed(z_mu_p)), list(reversed(z_var_p)), list(reversed(z_sample))
+
+    def sample_z(self, mu, var):
+        '''
+        Sampling z ~ p(z)= N(mu,var)
+        :return: sample of z
+        '''
+        stddev = var ** 0.5
+        out = mu + stddev * torch.randn(mu.size())
+        return out
+
+    def Gaussian_update(self, mu_q_d, var_q_d, mu_p, var_p):
+        '''
+        Combine mu_q_d,var_q_d and mu_p,var_p to generate mu_q,var_q
+        :return two tensor: mu_q,var_q
+        '''
+        x = torch.pow(var_q_d, -1)
+        y = torch.pow(var_p, -1)
+        sigma = torch.pow(torch.add(x, y), -1)
+        var = torch.pow(sigma, 2)
+        mu = torch.add(mu_q_d*x, mu_p*y) * sigma
+        return mu, var
```

Comment:
这个地方可能也是需要按照var实际上是log_var来考虑，即多加一步exp()的变换

Reply:
另外，文献中的式17应该是写错了，等式左边应该是方差而不是标准差，不然量纲不对，所以你这里也不应该再有sigma到var的一步
```suggestion
    def Gaussian_update(self, mu_q_d, var_q_d, mu_p, var_p):
        '''
        Combine mu_q_d,var_q_d and mu_p,var_p to generate mu_q,var_q
        :return two tensor: mu_q,var_q
        '''
        x = torch.pow(var_q_d, -1)
        y = torch.pow(var_p, -1)
        var = torch.pow(torch.add(x, y), -1)
        mu = torch.add(mu_q_d*x, mu_p*y) * sigma
        return mu, var
```